File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *l = head, *r = head;
        while (n > 0) {
            r = r->next;
            --n;
        }

        ListNode* prev = nullptr;
        while (r) {
            prev = l;
            l = l->next;
            r = r->next;
        }

        ListNode *result, *toDelete;
        if (!prev) {
            toDelete = head;
            result = toDelete->next;
        } else {
            toDelete = prev->next;
            prev->next = toDelete->next;
            result = head;
        }

        delete toDelete;
        return result;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-1.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun reorderList(head: ListNode?): Unit {
        if (head == null) { return }
        var slow = head
        var fast = head
        while (fast?.next?.next != null) {
            slow = slow?.next
            fast = fast?.next?.next
        }
        var second = slow?.next
        slow?.next = null
        second = reversed(second)
        var first = head
        while (second!=null) {
            val nextFirst = first?.next
            val nextSecond = second?.next
            first?.next = second
            second.next = nextFirst
            second = nextSecond
            first = nextFirst
        }
    }

    fun reversed(head: ListNode?): ListNode? {
        var prev: ListNode? = null
        var curr: ListNode? = head
        while (curr != null) {
            val next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        }
        return prev
    }
}
